src/main/kotlin/com/wgslfuzz/core/Resolver.kt

```diff
@@ -510,71 +557,87 @@ private fun nodeIntroducesNewScope(
             parentNode !is ContinuingStatement
     )
 
+private fun <T : Type> T.asStoreType(): Type =
+    when (this) {
+        is Type.Reference -> this.storeType
+        else -> this
+    }
+
+private fun defaultAccessModeOf(addressSpace: AddressSpace): AccessMode =
+    when (addressSpace) {
+        AddressSpace.FUNCTION -> AccessMode.READ_WRITE
+        AddressSpace.PRIVATE -> AccessMode.READ_WRITE
+        AddressSpace.WORKGROUP -> AccessMode.READ_WRITE
+        AddressSpace.UNIFORM -> AccessMode.READ
+        AddressSpace.STORAGE -> AccessMode.READ
+        AddressSpace.HANDLE -> AccessMode.READ
+    }
+
 private fun resolveExpressionType(
     expression: Expression,
     resolverState: ResolverState,
 ): Type =
     when (expression) {
         is Expression.FunctionCall -> resolveTypeOfFunctionCallExpression(expression, resolverState)
-        is Expression.IndexLookup ->
+        is Expression.IndexLookup -> {
```

Comment:
I think this probably does want extracting into its own helper, like `resolveTypeOfIndexLookupExpression`